File: express-mongoose-js/middlewares/error.middleware.js
import { HttpStatus } from '../utils/index.js';

export function errorMiddleware(err, _, res, __) {
  if (err) {
    let r = {
      status: err.statusText || 'UnknownError',
      message: err.message,
    };

    if (process.env.NODE_ENV === 'development') {
      r.stack = err.stack;
    }

    res.status(err.statusCode || HttpStatus.InternalServerError).json(r);
  }
}

File: gin-mongo-go/middlewares/error.middleware.go
package middlewares

import (
	"log"
	"net/http"
	"os"

	"github.com/gin-gonic/gin"

	"example.com/m/utils"
)

func ErrorMiddleware() gin.HandlerFunc {
	return func(c *gin.Context) {
		c.Next()

		if len(c.Errors) > 0 {
			err := c.Errors.Last().Err
			var env string
			if env = os.Getenv("APP_ENV"); env == "" {
				env = "production"
			}

			if e, ok := err.(utils.AppError); ok {
				response := gin.H{"message": e.Message}
				if env == "development" {
					response["details"] = e.Raw
				}
				c.JSON(e.Status, response)
				c.Abort()
				return
			}

			log.Printf("[CRITICAL ERROR] %v", err)
			response := gin.H{"message": "Something went wrong!"}
			if env == "development" {
				response["details"] = err.Error()
			}
			c.JSON(http.StatusInternalServerError, response)
		}

		c.Abort()
	}
}
